### PR TITLE
Add cram

### PR DESCRIPTION
**Pull request by fannyhb**
## Contents
Review deadline: 5/9

Main reviewer: @alvaralmstedt 

### The What
Adding cram compression to snakemake.

### The Why
We want to move from using bam to only using cram when transferring data to webstore.

### The How
Added new rule: cram.smk
Added new singularity samtools.sif
The singularity contains the reference genome and samtools v.1.20.

### This [update](https://semver.org/) is:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### Expected outcome
.cram and .cram.crai files should appear both in the workdir and BDC-portal.

## Verifications
- [ ] Code reviewed by @alvaralmstedt 
- [x] Code tested by @fannyhb

### CODE REVIEW

File: singularities/samtools/samtools-1.20.def [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Bootstrap: docker
+
+# The build-stage image:
+From: continuumio/miniconda3:24.4.0-0
+
+%files
+	/home/xberfa/scripts/build_sing/samtools-1.20.yml /dependencies/samtools-1.20.yml
```
Comment (JPWerling) [inline singularities/samtools/samtools-1.20.def:7]: No homely paths pretty please

File: singularities/samtools/samtools-1.20.def [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    /apps/bio/dependencies/wgs_somatic/hg38/sentieon/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna /references/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna
+    /apps/bio/dependencies/wgs_somatic/hg38/sentieon/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.fai /references/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.fai
```
Comment (JPWerling) [inline singularities/samtools/samtools-1.20.def:9]: Should these be here? Why are they not input arguments to the singularity when running the snakemake rule? Requires a bit of binding but that shouldn't be too difficult to manage

File: workflows/rules/mapping/cram.smk [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# coding: utf-8
+
+rule cram:
+    input:
+        bam = "{stype}/realign/{sname}_REALIGNED.bam",
+        bai = "{stype}/realign/{sname}_REALIGNED.bam.bai"
+    singularity:
+        pipeconfig["singularities"]["samtools"]["sing"]
+    params:
+        threads = clusterconf["mapping"]["threads"],
```
Comment (alvaralmstedt) [inline workflows/rules/mapping/cram.smk:11]: maybe compression can support more threads than mapping?

**Review by alvaralmstedt — APPROVED** [review]
Om du har testat, kör på :)

**Review by JPWerling — CHANGES_REQUESTED** [review]
Rethinking the way the samtools singularity is being built might be a good idea. An option to generate the image would be to run something akin to 
```
apptainer build samtools.img docker://staphb/samtools:1.20
```
`staphb` in my experience keeps softwares up to date fairly well. 

There is some value in managing the image building ourselves in the event that staphb decides to stop generating new images. However, us changing versions frequently is unlikely and so the image we generate now will likely live for a long time.

The only required change i see fit is regarding references in the image itself. Have the reference be an input to the container instead. In addition the conda yaml needs to be placed somewhere not in a users home.

File: singularities/samtools/samtools-1.20.def (hunk 1 above)
Comment (fannyhb) [reply singularities/samtools/samtools-1.20.def]: Removed the conda config.

File: singularities/samtools/samtools-1.20.def (hunk 2 above)
Comment (fannyhb) [reply singularities/samtools/samtools-1.20.def]: Removed, using the reference located on the cluster instead. Specified in the cluster config.

**Review by JPWerling — APPROVED** [review]
Looks good

**Pull request merged**